src/dmd/expression.d

```diff
@@ -4823,7 +4823,29 @@ extern (C++) final class DotVarExp : UnaExp
                                 if (!onlyUnion)
                                     v.ctorinit = false;
```

Comment:
So it looks like you just need to pass the `onlyUnion` information to `checkModify` so that it does not set `ctorinit` in the first place, possibly by extending the existing `flag` parameter.

Reply:
Currently, the initialization information is reverted in a specific case; you can see that there is a deep level of nesting here. All those checks would need to be reproduced in checkModify/modifyFieldVar. I think that this patch is better off this way.

Reply:
Not quite following that.  All that you've added here below looks to be duplicated from modifyFieldVar.

Reply:
They already exist in modifyFiel

> modifyFieldVar simply checks whether a variable is modifiable. In the context of the bug report, we have `this.var.field` and modifyFieldVar ends up being called on `var`. modifyFieldVar then simply does initialization checks on `var` and it considers it initialized regardless of the context. To be able to move this code in modifyFieldVar, we would need to check for the specific case of DotVarExp, essentially the conditions at lines 4811-4815 because otherwise we might end up wrongfully setting ctorinit for other cases.
> 

???

When `var.isField()` in a `modifyFieldVar` it can only come from a DotVarExp.